include/SFML/Window/Event.hpp

```diff
@@ -34,196 +34,298 @@
-        MouseButtonPressed,     //!< A mouse button was pressed (data in event.mouseButton)
-        MouseButtonReleased,    //!< A mouse button was released (data in event.mouseButton)
-        MouseMoved,             //!< The mouse cursor moved (data in event.mouseMove)
-        MouseEntered,           //!< The mouse cursor entered the area of the window (no data)
-        MouseLeft,              //!< The mouse cursor left the area of the window (no data)
-        JoystickButtonPressed,  //!< A joystick button was pressed (data in event.joystickButton)
-        JoystickButtonReleased, //!< A joystick button was released (data in event.joystickButton)
-        JoystickMoved,          //!< The joystick moved along an axis (data in event.joystickMove)
-        JoystickConnected,      //!< A joystick was connected (data in event.joystickConnect)
-        JoystickDisconnected,   //!< A joystick was disconnected (data in event.joystickConnect)
-        TouchBegan,             //!< A touch event began (data in event.touch)
-        TouchMoved,             //!< A touch moved (data in event.touch)
-        TouchEnded,             //!< A touch event ended (data in event.touch)
-        SensorChanged,          //!< A sensor value changed (data in event.sensor)
-
-        Count //!< Keep last -- the total number of event types
+        Empty,                  //!< Default type
+        Closed,                 //!< The window requested to be closed
+        Resized,                //!< The window was resized
+        LostFocus,              //!< The window lost the focus
+        GainedFocus,            //!< The window gained the focus
+        TextEntered,            //!< A character was entered
+        KeyPressed,             //!< A key was pressed
+        KeyReleased,            //!< A key was released
+        MouseWheelScrolled,     //!< The mouse wheel was scrolled
+        MouseButtonPressed,     //!< A mouse button was pressed
+        MouseButtonReleased,    //!< A mouse button was released
+        MouseMoved,             //!< The mouse cursor moved
+        MouseEntered,           //!< The mouse cursor entered the area of the window
+        MouseLeft,              //!< The mouse cursor left the area of the window
+        JoystickButtonPressed,  //!< A joystick button was pressed
+        JoystickButtonReleased, //!< A joystick button was released
+        JoystickMoved,          //!< The joystick moved along an axis
+        JoystickConnected,      //!< A joystick was connected
+        JoystickDisconnected,   //!< A joystick was disconnected
+        TouchBegan,             //!< A touch event began
+        TouchMoved,             //!< A touch moved
+        TouchEnded,             //!< A touch event ended
+        SensorChanged,          //!< A sensor value changed
     };
 
+    Event() = default;
+
+    template <typename T>
+    explicit Event(const T& t) : m_data(t)
```

Comment:
I know the `explicit` improves type safety here, but at the same time I also feel like being allowed to omit `sf::Event()` around a value feels more natural. I'm not fully up to date with template shenanigans, but I'm pretty sure there'd be some way more fancy approach for this.

Reply:
I was thinking the same thing. There's lots of `Event(event)` code that I didn't like writing.

The problem is that the template is currently unconstrained to removing `explicit` means the compiler will try to convert _anything_ into an Event. Kimci shared an idea for containing this template which I may look into.

Reply:
Yeah, I'd definitely use some constraints in some form or shape.